File: mobile/src/services/checkpoints.ts
// Checkpoints — the reward for clearing a whole skill, granted by the server.
//
// The mirror of frontend/services/checkpoints.ts, against the same endpoints, so
// a checkpoint collected in the browser is already paid when the app opens.
//
// The server owns the reward completely (backend/live-bridge/app/checkpoints.py):
// it decides WHICH skills are cleared by reading the learner's own state, it
// computes the XP from the authored step count, and it records the grant inside
// a Firestore transaction keyed on (uid, skillId). A second claim therefore
// grants zero, whatever this client does.
//
// What the CLIENT owns is the running XP total, which lives in the shared
// progress record. That split is the whole reason this file is careful: if the
// client also added XP optimistically, a learner who tapped twice, or collected
// on their phone and then opened the web, would be paid once by the server's
// record and twice by their own counter.
//
// So there is no optimistic add anywhere here. Instead the progress record
// carries a LEDGER, `checkpointXp`: how much checkpoint XP has already been
// folded into `xp`. Every fold names an ABSOLUTE total and pays only the
// difference, so running it twice, on two devices, out of order, or after the
// app is killed between the claim and the save all land on the same number.
// Nothing here ever adds a delta to whatever the ledger happens to hold: that
// is the one shape that double-pays, because two things can fold the same
// grant.

import { API_BASE } from './config';
import { getIdToken } from './firebase';

/** One cleared skill and what it paid. Returned by both endpoints. */
export interface CheckpointReward {
  skillId: string;
  title: string;
  unitId: string;
  xp: number;
}

export interface CheckpointStatus {
  /** skillId -> XP already granted. The server's record, not ours. */
  claimed: Record<string, number>;
  /** Cleared but unpaid: every one of these is collectable right now. */
  available: CheckpointReward[];
  /** Sum of `claimed`. The figure the local ledger reconciles against. */
  totalClaimedXp: number;
}

export interface CheckpointGrant {
  /** What this call actually paid. Empty means the claim was a no-op, which is
   *  the normal answer on a device that is catching up rather than earning. */
  granted: CheckpointReward[];
  xp: number;
}

/** Why a call failed, at the resolution the UI needs. Matches community.ts. */
export type FailReason =
  | 'offline'
  | 'timeout'
  | 'unauthenticated'
  | 'rate_limited'
  | 'server';

export type Result<T> = { ok: true; data: T } | { ok: false; reason: FailReason };

/** Anything carrying the XP total and the ledger beside it. */
export interface XpLedger {
  xp: number;
  /** Checkpoint XP already folded into `xp`. Absent on records written before
   *  checkpoints paid out, which is correctly read as zero: nothing had been
   *  granted then, because no client called the endpoint. */
  checkpointXp?: number;
}

/**
 * React Native's fetch has NO default timeout, so a request on a weak
 * connection can sit open indefinitely and the screen just looks dead. Twelve
 * seconds covers a Cloud Run cold start on a slow link and still surfaces a real
 * failure while the learner is looking at it.
 */
const TIMEOUT_MS = 12_000;

async function call<T>(path: string, init?: RequestInit): Promise<Result<T>> {
  if (!API_BASE) return { ok: false, reason: 'offline' };
  const token = await getIdToken();
  // A missing token is a session problem, not a network one, and telling
  // someone to check their connection when they need to sign in again wastes
  // their time.
  if (!token) return { ok: false, reason: 'unauthenticated' };

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
  try {
    const res = await fetch(`${API_BASE}/v1/me${path}`, {
      ...init,
      signal: controller.signal,
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json',
        ...(init?.headers ?? {}),
      },
    });
    if (res.status === 401 || res.status === 403) return { ok: false, reason: 'unauthenticated' };
    if (res.status === 429) return { ok: false, reason: 'rate_limited' };
    if (!res.ok) return { ok: false, reason: 'server' };
    return { ok: true, data: (await res.json()) as T };
  } catch (err) {
    const aborted = (err as { name?: string } | null)?.name === 'AbortError';
    return { ok: false, reason: aborted ? 'timeout' : 'offline' };
  } finally {
    clearTimeout(timer);
  }
}

/** What has been granted, and what is earned and waiting. */
export function fetchCheckpoints(): Promise<Result<CheckpointStatus>> {
  return call<CheckpointStatus>('/checkpoints');
}

// One claim in flight at a time, so a second tap costs no second round trip.
// Note what this dedupe does NOT do: both callers get the SAME reply, so both
// see the same non-empty grant, and both fold it. Correctness therefore cannot
// rest here or on the button's disabled state — it rests on `foldClaim` naming
// an absolute total, which makes the second fold a no-op.
let claiming: Promise<Result<CheckpointGrant>> | null = null;

/**
 * Collect every checkpoint earned and not yet paid for.
 *
 * Never retried. A POST that timed out may well have been applied, and the reply
 * carries the ceremony: a retry could celebrate a grant that already happened,
 * or show nothing for one that did.
 */
export function claimCheckpoints(): Promise<Result<CheckpointGrant>> {
  if (claiming) return claiming;
  const p = call<CheckpointGrant>('/checkpoints/claim', { method: 'POST' });
  claiming = p;
  // Cleared from outside the promise chain, so a synchronous rejection cannot
  // latch this forever.
  void p.finally(() => { if (claiming === p) claiming = null; });
  return p;
}

/**
 * Fold the server's total granted XP into a progress record, exactly once.
 *
 * Returns the patch to apply, or null when the record is already square with
 * the server, which is the common case and worth detecting so an untouched
 * record is not rewritten on every visit.
 *
 * `checkpointXp` is a HIGH-WATER MARK, not simply the last total seen. A total
 * LOWER than the mark can only be a stale read: a GET issued before a claim
 * committed and answered after it. Writing that back would leave the granted XP
 * sitting inside `xp` with nothing accounting for it, and the next fetch would
 * pay it a second time. A genuinely lower server total means the claim record
 * was restored from a backup; holding the mark under-pays a re-grant in that one
 * case, which is the right way round to be wrong.
 */
export function foldCheckpointXp<T extends XpLedger>(
  progress: T,
  totalClaimedXp: number,
): Pick<XpLedger, 'xp' | 'checkpointXp'> | null {
  if (!Number.isFinite(totalClaimedXp) || totalClaimedXp < 0) return null;
  const folded = progress.checkpointXp ?? 0;
  if (totalClaimedXp <= folded) return null;
  return { xp: progress.xp + (totalClaimedXp - folded), checkpointXp: totalClaimedXp };
}

/**
 * Fold a claim this client has just been handed, without waiting to re-read the
 * server's total.
 *
 * `totalClaimedBefore` is the total this client last saw from GET /checkpoints,
 * captured when the claim was fired. With `grantedXp` it names an ABSOLUTE
 * total, and that is what makes this safe. Adding a delta to whatever the ledger
 * happens to hold when the reply lands double-pays whenever something else
 * folded the same grant first, and two things can: a background status refresh
 * that raced the claim, and a second caller awaiting the same deduplicated
 * promise. Folding to a fixed number makes both of those no-ops, because the
 * ledger is already standing there.
 *
 * It can never land long. The skills in `granted` were absent from the claim
 * record the server read, so they were absent from `totalClaimedBefore` too, and
 * that record only grows: `totalClaimedBefore + grantedXp` is at most the
 * server's true total. When it lands short, because another device collected
 * something this record never saw, the next `foldCheckpointXp` against a
 * fetched status adds the remainder.
 */
export function foldClaim<T extends XpLedger>(
  progress: T,
  totalClaimedBefore: number,
  grantedXp: number,
): Pick<XpLedger, 'xp' | 'checkpointXp'> | null {
  if (!Number.isFinite(grantedXp) || grantedXp <= 0) return null;
  if (!Number.isFinite(totalClaimedBefore) || totalClaimedBefore < 0) return null;
  return foldCheckpointXp(progress, totalClaimedBefore + grantedXp);
}

File: backend/live-bridge/app/checkpoints.py
"""Checkpoints — the reward for clearing a skill, granted once and only once.

A unit holds 11-14 lessons across 4-5 skills. The path draws a checkpoint at
every skill boundary; this is what makes reaching one worth something.

Three properties matter, and all three are why this is server work rather than a
client counter:

  1. **The amount is computed here.** It scales with the skill's lesson count
     (15 XP each, so a two-lesson skill pays 30 and a four-lesson skill 60),
     read from the authored curriculum. A client that asked for its own number
     could ask for any number.

  2. **The claim is verified.** Before granting anything we read the learner's
     own state envelope and check that every lesson in the skill is actually
     complete. "I finished this" is a claim, not a fact.

  3. **The claim is atomic.** Granted inside a Firestore transaction keyed on
     (uid, skillId). Without it, replaying a skill — or two devices claiming at
     once — is an XP printing press. This is the same failure hearts would have
     had with a device-owned balance, and it gets the same answer.

XP itself still lives in the client's progress envelope, so the grant is returned
rather than written: the server decides IF and HOW MUCH, the client records it.
The claim record is what makes that safe, because a second attempt grants zero no
matter what the client does.
"""

from __future__ import annotations

import logging
import os
from datetime import datetime, timezone
from functools import lru_cache
from typing import Any

from google.cloud import firestore

import curriculum

logger = logging.getLogger("ohmlet.checkpoints")

CHECKPOINTS_COLLECTION = os.getenv("OHMLET_CHECKPOINTS_COLLECTION", "ohmlet_checkpoints")
STATE_COLLECTION = os.getenv("OHMLET_STATE_COLLECTION", "ohmlet_state")

# Per STEP in the skill, not per lesson.
#
# Per-lesson was packaging-dependent, and the packaging changed: splitting the
# authored lessons into shorter sessions doubled every skill's lesson count
# without adding a single question, and checkpoint payouts doubled with it —
# from 22% of total lesson XP to 48%, purely from a delivery decision. Steps are
# the invariant: the same work pays the same bonus however it is sliced.
#
# At 0.5 the checkpoints pay ~1,180 against the lessons' ~4,700, about 25% on
# top. The ceremony is the reward; the XP is the signal.
XP_PER_STEP = float(os.getenv("OHMLET_CHECKPOINT_XP_PER_STEP", "0.5"))

# A skill with a single lesson gets no checkpoint. Fifteen of the 57 skills hold
# exactly one lesson, and a boundary marked after every one of them would fire a
# celebration every other screen — which is how a reward stops being one. Only
# the 42 multi-lesson skills carry a checkpoint, roughly one per three lessons.
MIN_LESSONS_FOR_CHECKPOINT = int(os.getenv("OHMLET_CHECKPOINT_MIN_LESSONS", "2"))


@lru_cache(maxsize=1)
def _skill_index() -> dict[str, dict[str, Any]]:
    """skillId -> {unitId, title, lessonIds}, for skills that CARRY a checkpoint.

    Single-lesson skills are absent, so they can never be claimed and never
    appear as available — the eligibility rule lives in one place rather than
    being re-decided at each call site."""
    index: dict[str, dict[str, Any]] = {}
    data = curriculum._curriculum()
    store = curriculum._lessons().get("lessons", {})
    for unit in data.get("units", []):
        for skill in unit.get("skills", []):
            sid = skill.get("id")
            if not sid:
                continue
            lesson_ids = [l.get("id") for l in skill.get("lessons", []) if l.get("id")]
            if len(lesson_ids) < MIN_LESSONS_FOR_CHECKPOINT:
                continue
            index[sid] = {
                "unitId": unit.get("id"),
                "title": skill.get("title", ""),
                "lessonIds": lesson_ids,
                "steps": sum(len(store.get(lid, {}).get("steps", [])) for lid in lesson_ids),
            }
    return index


def xp_for(skill_id: str) -> int:
    entry = _skill_index().get(skill_id)
    if not entry:
        return 0
    # Rounded to a multiple of 5 so the number on the card reads as a reward
    # rather than as arithmetic.
    return max(5, round(entry["steps"] * XP_PER_STEP / 5) * 5)


def _completed_lessons(client: firestore.Client, uid: str) -> set[str]:
    """Lesson ids the learner has finished, read from their state envelope.

    Absent or malformed state means an empty set, which grants nothing. Failing
    closed is right here: the cost is a checkpoint claimed a moment later, and
    the alternative is granting XP for work we cannot see.
    """
    try:
        snap = client.collection(STATE_COLLECTION).document(uid).get()
        if not snap.exists:
            return set()
        data = (snap.to_dict() or {}).get("data") or {}
        levels = data.get("lessonLevels")
        return set(levels.keys()) if isinstance(levels, dict) else set()
    except Exception as exc:
        logger.warning("state read failed for %s: %s", uid, exc)
        return set()


def _claimed(client: firestore.Client, uid: str) -> dict[str, int]:
    try:
        snap = client.collection(CHECKPOINTS_COLLECTION).document(uid).get()
        if snap.exists:
            got = (snap.to_dict() or {}).get("claimed")
            if isinstance(got, dict):
                return {k: int(v.get("xp", 0)) for k, v in got.items() if isinstance(v, dict)}
    except Exception as exc:
        logger.warning("checkpoint read failed for %s: %s", uid, exc)
    return {}


def status(uid: str) -> dict[str, Any]:
    """What the learner has claimed, and what is sitting there unclaimed."""
    from state_store import get_client

    client = get_client()
    claimed = _claimed(client, uid)
    done = _completed_lessons(client, uid)
    available = [
        {"skillId": sid, "title": entry["title"], "unitId": entry["unitId"], "xp": xp_for(sid)}
        for sid, entry in _skill_index().items()
        if sid not in claimed and entry["lessonIds"] and done.issuperset(entry["lessonIds"])
    ]
    return {
        "claimed": claimed,
        "available": available,
        "totalClaimedXp": sum(claimed.values()),
    }


def claim_all(uid: str) -> dict[str, Any]:
    """Grant every checkpoint the learner has earned and not yet been paid for.

    Claiming everything available in one call rather than one skill at a time is
    deliberate: it is naturally idempotent (a second call finds nothing left),
    it survives a client that missed a claim while offline, and it costs one
    round trip instead of one per skill.
    """
    from state_store import get_client

    client = get_client()
    done = _completed_lessons(client, uid)
    index = _skill_index()
    ref = client.collection(CHECKPOINTS_COLLECTION).document(uid)

    @firestore.transactional
    def _grant(txn: firestore.Transaction) -> list[dict[str, Any]]:
        snap = ref.get(transaction=txn)
        existing = (snap.to_dict() or {}).get("claimed") if snap.exists else {}
        existing = existing if isinstance(existing, dict) else {}

        granted: list[dict[str, Any]] = []
        writes: dict[str, Any] = {}
        now = datetime.now(timezone.utc).isoformat()
        for sid, entry in index.items():
            if sid in existing or not entry["lessonIds"]:
                continue
            if not done.issuperset(entry["lessonIds"]):
                continue
            xp = xp_for(sid)
            writes[f"claimed.{sid}"] = {"xp": xp, "at": now}
            granted.append({"skillId": sid, "title": entry["title"], "unitId": entry["unitId"], "xp": xp})

        if writes:
            # set(merge) rather than update(): the document may not exist yet, and
            # update() would raise on a first-ever claim.
            payload: dict[str, Any] = {"uid": uid, "claimed": {}}
            for key, value in writes.items():
                payload["claimed"][key.split(".", 1)[1]] = value
            txn.set(ref, payload, merge=True)
        return granted

    try:
        granted = _grant(client.transaction())
    except Exception as exc:
        # Loud, not silent: a swallowed failure here looks to the learner like a
        # checkpoint that paid nothing, and they would have no way to tell.
        logger.error("checkpoint claim failed for %s: %s", uid, exc)
        raise

    return {
        "granted": granted,
        "xp": sum(g["xp"] for g in granted),
    }
